### PR TITLE
Hardcode changes for Apple Silicon support

### LINKED ISSUE #4664: Rust parser doesn't link/run on Mac M1
This task is automatically imported from the old Task Issue Board and it was originally created by jaroslavtulach.
Original issue is [here](https://www.pivotaltracker.com/story/show/183635511).
<hr />

Marcin [has discovered](https://discord.com/channels/401396655599124480/1034429857431556107) that the `libenso_parser.dylib` cannot be loaded into the Enso engine on his Mac M1. Very likely the library isn't created for the right architecture.

GraalVM doesn't support M1 processor (tracked as #182269744) and runs in rosetta emulator. Rust builds for ARM by default, so if Graal runs on Rosetta it's not going to work. We need to fix the build script to x-compile.

## Comments:
I have found [following advice](https://stackoverflow.com/questions/68139162/how-can-i-cross-compile-rust-code-into-intel-assembly-on-an-arm-m1-apple-silicon) on stackoverflow. (jaroslavtulach - Oct 26, 2022)
<hr />
It works with the following changes (though this diff would break any other platform, so some love is needed):

```diff
diff --git a/build.sbt b/build.sbt
index 3430e0c82..49d2fdc40 100644
--- a/build.sbt
+++ b/build.sbt
@@ -678,7 +678,7 @@ val generateRustParserLib =
     allLibs.isEmpty ||
     (`syntax-rust-definition` / generateRustParserLib).inputFileChanges.hasChanges
   ) {
-    Seq("cargo", "build", "-p", "enso-parser-jni") !
+    Seq("cargo", "build", "-p", "enso-parser-jni", "--target", "x86_64-apple-darwin" ) !
   }
   FileTreeView.default.list(Seq(libGlob)).map(_._1.toFile)
 }

diff --git a/lib/rust/parser/generate-java/java/org/enso/syntax2/Parser.java b/lib/rust/parser/generate-java/java/org/enso/syntax2/Parser.java
index cd4b3be7e..c0b3731a6 100644
--- a/lib/rust/parser/generate-java/java/org/enso/syntax2/Parser.java
+++ b/lib/rust/parser/generate-java/java/org/enso/syntax2/Parser.java
@@ -13,7 +13,7 @@ public final class Parser implements AutoCloseable {
     for (; ; ) {
       File parser;
       if (os.startsWith("Mac")) {
-        parser = new File(dir, "target/rust/debug/libenso_parser.dylib");
+        parser = new File(dir, "target/rust/x86_64-apple-darwin/debug/libenso_parser.dylib");
       } else if (os.startsWith("Windows")) {
         parser = new File(dir, "target/rust/debug/enso_parser.dll");
       } else {
```

The fix is _delivered_.

Marcin got everything working on his Mac M1. I integrated the changes as part of [PR-3842](https://github.com/enso-org/enso/pull/3842). However later @murbanczyk found we have problems on regular Macs. As such I had to integrate also [PR-3850](https://github.com/enso-org/enso/pull/3850). That might again break Mac M1, alas.

Marking _delivered_. Giving Marcin (as the only known owner of Mac M1) a chance to _accept_/_reject_. (jaroslavtulach - Nov 4, 2022)
<hr />

### PR DESCRIPTION
**Pull request by JPMoresmau**
### Pull Request Description

As requested in https://github.com/enso-org/enso/issues/4664#issuecomment-1563863054, these are the changes I made to be able to build Enso on Apple Silicon.

### Important Notes

These changes do not try to be cross-platform compatible and WILL BREAK the build on all other platforms. @wdanilo gave me permission :-). This is just to show what changes are required to build on M1. 

### Checklist

N/A

### CODE REVIEW

File: lib/rust/parser/generate-java/java/org/enso/syntax2/Parser.java [hunk 1]
```diff
@@ -26,7 +26,7 @@ public final class Parser implements AutoCloseable {
       System.load(parser.getAbsolutePath());
     } catch (URISyntaxException | LinkageError e) {
       File root = new File(".").getAbsoluteFile();
-      if (!searchFromDirToTop(e, root, "target", "rust", "debug", name)) {
+      if (!searchFromDirToTop(e, root, "target", "rust", "x86_64-apple-darwin", "debug", name)) {
```
Comment (JaroslavTulach) [inline lib/rust/parser/generate-java/java/org/enso/syntax2/Parser.java]: This is reverting [Marcin's Nov 9 change](https://github.com/enso-org/enso/commit/23e04f905f89b857ecd58a733f1aad4a63338513#diff-e237fc34931809c65628b8bcde58e32add8959b8e66940c46a5f63f557530302L30) done as part of #3859 - don't you have some old Rust tool chain?

Comment (JPMoresmau) [reply lib/rust/parser/generate-java/java/org/enso/syntax2/Parser.java]: Rustup says it's 1.26...

Comment (wdanilo) [reply lib/rust/parser/generate-java/java/org/enso/syntax2/Parser.java]: This is very old! We are using 1.68 now. 1.26 was released May 10, 2018.

Comment (JPMoresmau) [reply lib/rust/parser/generate-java/java/org/enso/syntax2/Parser.java]: 1.26 is the rustup version, which I thought was the question, and the latest version. 
```
rustup 1.26.0 (5af9b9484 2023-04-05)
info: This is the version for the rustup toolchain manager, not the rustc compiler.
info: The currently active `rustc` version is `rustc 1.68.0-nightly (1e4f90061 2023-01-11)`
```

Comment (wdanilo) [reply lib/rust/parser/generate-java/java/org/enso/syntax2/Parser.java]: Rustup is not Rust toolchain, it is Rust toolchain installer :) 

This is a little worrying that a revert of Marcins commit is needed here. @JaroslavTulach would you be able to ask Marcin about it, please? Maybe he would remember why he did it.

Comment (JPMoresmau) [reply lib/rust/parser/generate-java/java/org/enso/syntax2/Parser.java]: Sorry for the confusion. I've undone that change locally and I'm trying to run clean / build and it seems to work, so maybe even though at some stage I thought I needed that extra folder in the path in fact it's not needed. I'm still not too sure how the full build chain works...

Comment (wdanilo) [reply lib/rust/parser/generate-java/java/org/enso/syntax2/Parser.java]: Ok, perfect. Thanks so much for the update! And thanks @JaroslavTulach for catching it!

**Comment by JaroslavTulach** [comment]
> ask Marcin about it

@kustosz is currently sailing in Jadran. He should be back next week. Let's wait for him. A QA by some additional M1 owner is needed anyway.

**Comment by JaroslavTulach** [comment]
There is a [formatting error](https://github.com/enso-org/enso/actions/runs/5089638064/jobs/9148896203?pr=6862#step:5:135) - running `sbt scalafmtSbt` shall fix it.

**Comment by JPMoresmau** [comment]
> There is a [formatting error](https://github.com/enso-org/enso/actions/runs/5089638064/jobs/9148896203?pr=6862#step:5:135) - running `sbt scalafmtSbt` shall fix it.

Sorry, didn't know about the formatting, should now be fixed.

**Comment by JaroslavTulach** [comment]
@kustosz will take a look next week.

**Comment by kustosz** [comment]
Confirming this worked for me :)

**Comment by JaroslavTulach** [comment]
> Confirming this worked for me :)

OK. So now we need to deal with [failure on regular systems](https://github.com/enso-org/enso/actions/runs/5090424644/jobs/9149448042?pr=6862#step:10:787).

**Comment by indiv0** [comment]
Potentially relevant prior attempt at this: https://github.com/enso-org/enso/pull/4102/files

<blockquote><div> GitHub</div><div><strong><a href="https://github.com/enso-org/enso/pull/4102">Fix M1 macOS path mismatch by mwu-tow · Pull Request #4102 · enso-org/enso</a></strong></div><div>Pull Request Description
This attempts to fix path mismatch in the build script, when the local architecture and architecture targeted by Graal differ.
Important Notes

Checklist
Please include the...</div></blockquote>

**Comment by xvcgreg** [comment]
@mwu-tow @JPMoresmau What is the status here?

**Comment by mwu-tow** [comment]
> @mwu-tow @JPMoresmau What is the status here?

This PR breaks builds on other platforms.  Information from here should be used to complete #4102, which is missing on the parser bits (as the new parser was merged after I prepared that PR).

Also, at this point, it might be worth to consider building native aarch64 images of the backend executables, I think that now GraalVM added support for this. Then likely the whole thing could be significantly simpler.

**Comment by xvcgreg** [comment]
@JaroslavTulach please check this GraalVM support with @mwu-tow and what would be the next steps?

**Comment by JaroslavTulach** [comment]
>> @mwu-tow @JPMoresmau What is the status here?
> This PR breaks builds on other platforms

Right. Until the gates are green, this PR cannot be integrated. I believe it is a task for @JPMoresmau (as the submitter of this PR) to make sure all the gates are green.

> Also, at this point, it might be worth to consider building native aarch64 images of the backend executables, 

Great idea!

> what would be the next steps?

- setup a builder to run on M1 Mac
- download GraalVM aarch64 bits instead of x64 bits
- configure parser to produce aarch64 `dylib`

At least those are the steps I envision. I have no way to verify or help to implement such vision, as I don't have access to M1 Mac.

**Comment by xvcgreg** [comment]
@JPMoresmau how would you like to approach this? Are you interested in fixing this issue?

**Comment by xvcgreg** [comment]
Converted to draft due to the fact this PR was not ready to merge.

**Comment by hubertp** [comment]
After GraalVM upgrade and #8407 #8471 this has become obsolete.

**Pull request closed without merging**